### ISSUE TITLE
Tracking issue for `IntoFuture`

### ISSUE THREAD

**Issue by seanmonstar** [rust-lang/rust#67644]
Introduced in https://github.com/rust-lang/rust/pull/65244, this is a trait of a type that can be converted into a `Future`, used by `.await` (similar to how `IntoIterator` is used with `for`).

*a more complete description can be found in https://github.com/rust-lang/rust/issues/67644#issuecomment-986257951*

**Comment by Fishrock123** [comment]
Are there any blockers on this?

[`Surf`](https://github.com/http-rs/surf) would benefit from this, as presently, [we use a kinda hacky work around with `Option`s](https://github.com/http-rs/surf/blob/c66f051591a3375f85e2184b2c12d2c307ddf667/src/request_builder.rs#L279-L290).

**Comment by Fishrock123** [comment]
cc @yoshuawuyts who may have some context

**Comment by Aaron1011** [comment]
This was implemented, but then reverted due to a performance regression. https://github.com/rust-lang/rust/pull/69218 may help reduce the impact of implementing this.

**Comment by kushwahashiv** [comment]
how to use this in an app if I want to still use it

**Comment by kushwahashiv** [comment]
any update?

**Comment by ibraheemdev** [comment]
Trait was added back in #72459 and `.await` integration is tracking in #67982

**Comment by WaffleLapkin** [comment]
Just for a clarification (and to be similar to newer tracking issues)

---

Feature gate: `#![feature(into_future)]`

This is a tracking issue for the `IntoFuture` trait similar to `IntoIterator` allowing to convert types into futures.

Since #90737 this trait is also automatically used in `expr.await` desugaring allowing for code similar to the following [(playground)](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a6bd2d1b0ff82ae75c81bb00918c770f):

```rust
struct NotAFutureYet;

impl IntoFuture for NotAFutureYet {
    type Output = u8;
    type Future = Ready<Self::Output>;
    
    fn into_future(self) -> Self::Future {
        ready(42)
    }
}

async fn test() {
    assert_eq!(NotAFutureYet.await, 42);
}
```

### Public API

```rust
// core::future

pub trait IntoFuture {
    type Output;
    type Future: Future<Output = Self::Output>;

    fn into_future(self) -> Self::Future;
}
```

### Steps / History

- [x] Implementation: #72459, #90737
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by WaffleLapkin** [comment]
I wonder though if the method should be called `into_fut` to be more consistent with `into_iter` (note that it's not called `into_iterator`)

**Comment by yoshuawuyts** [comment]
> I wonder though if the method should be called `into_fut` to be more consistent with `into_iter` (note that it's not called `into_iterator`)

@WaffleLapkin I don't think we should; "future" is a pretty short name already. The "iterator" submodule is named `core::iter` and the "future" submodule is named `core::future`. Even if we shorten "future" to "fut" for this API, we wouldn't be able to consistently match how it's used for iterator. I feel that inconsistency might throw people off, and it'd be better if we keep using "future" everywhere.

---

@rust-lang/libs-api now that https://github.com/rust-lang/rust/issues/67982 has landed `.await` now desugars using `IntoFuture` as per [RFC 2394](https://rust-lang.github.io/rfcs/2394-async_await.html#the-expansion-of-await). This means two years after the current version of `IntoFuture` [landed on nightly](https://github.com/rust-lang/rust/pull/72459) the last blocker for it has been resolved!

I'd like to propose we FCP `IntoFuture` for stabilization. Rust 1.59 will be using the new `.await` desugaring, and it'd be great if the trait it desugars into would also be available on stable.

**Comment by dtolnay** [comment]
:-1: I'd veto stabilizing `IntoFuture` in the same release as the desugaring change. I don't know of any technical reason that the desugaring and trait benefit from being stabilized in the same release. I know the trait is old, but it hasn't been used in `.await` so far. Let's give users a chance to write code against this on nightly and uncover any resulting API awkwardness. The desugaring change can be amended or rolled back at any point because `impl<F: Future> IntoFuture for F` is the only impl built into core that a stable `.await` could make use of.

**Comment by yoshuawuyts** [comment]
I just realized that the associated type of the `IntoFuture` trait is `Future`, while the associated type of `IntoIterator` is `IntoIter`. It would probably make sense to keep the two consistent and rename the associated type to `IntoFuture`. Not sure why I missed that in https://github.com/rust-lang/rust/pull/72459, but it's not yet too late to correct it.

**Comment by yoshuawuyts** [comment]
Now that Rust 1.59 with the `IntoFuture` desugaring changes is live, I think we're getting close to being able to stabilize the trait too. In order to assess how it would affect a real-world codebase I've run an assessment by converting 31 APIs in the Azure SDK to rely on `IntoFuture` desugaring: https://github.com/Azure/azure-sdk-for-rust/pull/687.

**tldr:** All but one API could be converted to use the `IntoFuture` trait, and the only one we couldn't convert was because of a lack of GATs.

---

In the last Async WG spring planning meeting @tmandry mentioned the following:

> __tmandry__: [at]Yoshua Wuyts [he/they] btw I remember seeing something that made me think that change still had a significant impact on the size of some futures, but I don't have a reference right now. I doubt it should impact stabilization
>
> __tmandry__: basically we had to bump the stack size of some tests

It'd be helpful if we could get a reference for that. As mentioned earlier in the thread, a several percentage-point increase in codegen size for `IntoFuture` desugaring is expected since we're doing more work. But we should verify that it isn't the type of blow-up that would warrant reverting the `IntoFuture` desugaring.

Once we have confirmation of that, I think we should be able to move ahead to propose stabilization!

**Comment by tmandry** [comment]
The regression was in stack size (i.e. the generator objects got bigger). Reference: https://fuchsia-review.googlesource.com/c/fuchsia/+/615168

The good news is that we determined that the regression only occurred in debug builds; release builds were not affected. The debug build regression on that test was significant, however; the required stack size went from around 1.5MiB to 2.5MiB.

This was needed when bumping rustc from 48a5999fceeea84a8971634355287faa349909d4...940a97a91023d1817b33d0ababc985793c9ed780, which is a much larger range than most bumps and so not very useful. However, I believe the regression occurred after https://github.com/rust-lang/rust/pull/72459.

Since this only seems to be a debug regression and it's already "priced in" I don't think this needs to block stabilization.

> As mentioned earlier in the thread, a several percentage-point increase in codegen size for `IntoFuture` desugaring is expected since we're doing more work.

My understanding is that `IntoFuture` should be fully transparent at runtime for types that are already futures. There might be some optimization quirks that prevent this, but I'd consider them bugs.

As a side note, I want to point out that several percentage points in production builds _can_ be considered a large regression for resource-constrained environments. I don't think that applies here, but I'd like to stay mindful of those use cases.

**Comment by yoshuawuyts** [comment]
> Since this only seems to be a debug regression and it's already "priced in" I don't think this needs to block stabilization.

That's great to hear, thanks for sharing!

@rust-lang/libs I would like to propose we stabilize this for 1.61.

**Comment by yaahc** [comment]
> @rust-lang/libs I would like to propose we stabilize this for 1.61. Can we start an FCP on this?

sounds good to me, though fwiw the correct team for reviewing this is @rust-lang/libs-api not libs. Also, we're in the process of updating our docs on the stabilization process and FCPS[^1] so if you wanted to follow the new process and start off the FCP with a stabilization report that would be greatly appreciated. It looks like your earlier comment is already essentially an experience report and @WaffleLapkin's comment already includes the API summary which I'm assuming is up to date so those sections could just link directly to those two comments, but I'd love to also see an implementation history.

**Comment by tmandry** [comment]
Another note on stack size: I thought adding `#[inline(always)]` to the [blanket impl](https://github.com/rust-lang/rust/blob/4ca19e09d302a4cbde14f9cb1bc109179dc824cd/library/core/src/future/into_future.rs#L25) might improve things for debug builds, but in my testing it did not seem to improve stack usage at all.

**Comment by yaahc** [comment]
Fantastic report, thank you

@rfcbot merge

**Comment by dtolnay** [comment]
Both is correct given that this affects meaning of `.await` language syntax.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

This is great, looking forward to this.

**Comment by yoshuawuyts** [comment]
ping @Amanieu, @cramertj, @m-ou-se - can I ask you to take a look at this when you have time?

**Comment by yaahc** [comment]
> I'm guessing this should be tagged with relnotes, or does to-announce already imply that?

:+1:, David went and tagged the PR
